kernel/thread.c

```diff
@@ -167,61 +178,109 @@ void z_thread_monitor_exit(struct k_thread *thread)
 }
 #endif
 
-#ifdef CONFIG_THREAD_NAME
-void z_impl_k_thread_name_set(struct k_thread *thread, const char *value)
+int z_impl_k_thread_name_set(struct k_thread *thread, const char *value)
 {
+#ifdef CONFIG_THREAD_NAME
 	if (thread == NULL) {
-		_current->name = value;
-	} else {
-		thread->name = value;
+		thread = _current;
 	}
+
+	strncpy(thread->name, value, CONFIG_THREAD_MAX_NAME_LEN);
+	thread->name[CONFIG_THREAD_MAX_NAME_LEN - 1] = '\0';
+	return 0;
+#else
+	ARG_UNUSED(thread);
+	ARG_UNUSED(value);
+	return -ENOSYS;
+#endif /* CONFIG_THREAD_NAME */
 }
 
-const char *z_impl_k_thread_name_get(struct k_thread *thread)
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(k_thread_name_set, thread, str_param)
 {
-	return (const char *)thread->name;
-}
+#ifdef CONFIG_THREAD_NAME
+	struct k_thread *t = (struct k_thread *)thread;
+	size_t len;
+	int err;
+	const char *str = (const char *)str_param;
+
+	if (t != NULL) {
+		if (Z_SYSCALL_OBJ(t, K_OBJ_THREAD) != 0) {
+			return -EINVAL;
+		}
+	}
+
+	len = z_user_string_nlen(str, CONFIG_THREAD_MAX_NAME_LEN, &err);
+	if (err != 0) {
+		return -EFAULT;
+	}
+	if (Z_SYSCALL_MEMORY_READ(str, len + 1) != 0) {
+		return -EFAULT;
+	}
 
+	return z_impl_k_thread_name_set(t, str);
 #else
-void z_impl_k_thread_name_set(k_tid_t thread_id, const char *value)
-{
-	ARG_UNUSED(thread_id);
-	ARG_UNUSED(value);
+	return -ENOSYS;
+#endif /* CONFIG_THREAD_NAME */
 }
+#endif /* CONFIG_USERSPACE */
 
-const char *z_impl_k_thread_name_get(k_tid_t thread_id)
+const char *k_thread_name_get(struct k_thread *thread)
 {
-	ARG_UNUSED(thread_id);
+#ifdef CONFIG_THREAD_NAME
+	return (const char *)thread->name;
+#else
+	ARG_UNUSED(thread);
 	return NULL;
-}
 #endif /* CONFIG_THREAD_NAME */
+}
```

Comment:
Same comment.   Just doesn't seem useful to have an API return a runtime failure when we already have a kconfig to disable the whole feature.

Reply:
See above, was already like this, I don't disagree but should be in another PR.